File: csharp/examples/GraduationVerdict/AtLeastNRule.cs
using VerdictRules;

namespace GraduationVerdict;

/// <summary>
/// Passes if at least <paramref name="minimum"/> of the given sub-rules pass.
/// </summary>
/// <remarks>
/// Not part of verdict-rules itself; see docs/extending/new-rule-shape/.
/// Evaluates every sub-rule unconditionally.
/// </remarks>
public sealed class AtLeastNRule(string name, IReadOnlyList<IRule> rules, int minimum, string? group = null) : IRule
{
    public string Name { get; } = name;

    public string? Group { get; } = group;

    public async Task<RuleResult> EvaluateAsync(
        IReadOnlyDictionary<string, object?> context,
        CancellationToken cancellationToken = default)
    {
        var subResults = new List<RuleResult>();
        foreach (var rule in rules)
        {
            subResults.Add(await rule.EvaluateAsync(context, cancellationToken));
        }
        var passedCount = subResults.Count(r => r.Passed);
        return new RuleResult(
            Name,
            passedCount >= minimum,
            $"{passedCount} of {rules.Count} passed, needed {minimum}",
            subResults);
    }
}

File: dart/examples/graduation_verdict/lib/src/at_least_n_rule.dart
import 'package:verdict_rules/verdict_rules.dart';

/// Passes if at least [minimum] of the given sub-rules pass.
///
/// Same shape as verdict_rules' docs/extending/new-rule-shape/
/// (`ThresholdRule`) -- not part of verdict_rules itself, a
/// consumer-defined combinator for a requirement [AndRule]/[OrRule] can't
/// express directly. Evaluates every sub-rule unconditionally (no
/// short-circuit is possible for a threshold count), unlike
/// [AndRule]/[OrRule].
class AtLeastNRule implements Rule<Context> {
  @override
  final String name;

  @override
  final String? group;

  final List<Rule<Context>> _rules;
  final int _minimum;

  AtLeastNRule(this.name, List<Rule<Context>> rules, int minimum, {this.group})
      : _rules = rules,
        _minimum = minimum;

  @override
  Future<RuleResult> evaluate(Context context) async {
    final subResults = <RuleResult>[];
    for (final rule in _rules) {
      subResults.add(await rule.evaluate(context));
    }
    final passedCount = subResults.where((r) => r.passed).length;
    return RuleResult(
      ruleName: name,
      passed: passedCount >= _minimum,
      detail: '$passedCount of ${_rules.length} passed, needed $_minimum',
      data: subResults,
    );
  }
}
